kedro-airflow/kedro_airflow/plugin.py

```diff
@@ -206,22 +205,19 @@ def create(  # noqa: PLR0913, PLR0912
         if group_in_memory:
             nodes, dependencies = group_memory_nodes(context.catalog, pipeline)
         else:
-            nodes = {node.name: [node] for node in pipeline.nodes}
-
-            dependencies = defaultdict(list)
+            nodes = {}
+            dependencies = {}
+            for node in pipeline.nodes:
+                nodes[node.name] = [node]
+                dependencies[node.name] = []
             for node, parent_nodes in pipeline.node_dependencies.items():
```

Comment:
Why do you believe that this row will be deterministic? I mean it affects the order of `.append(node.name)`

Reply:
We are iterating nodes in the order they were resolved with topological sort, which is deterministic, and `nodes` and `dependencies` dictionaries are ordered.

Reply:
`dependencies[parent.name].append(node.name)`
I understand that the parents will be deterministic because you initialised an empty dictionary previously. However, could you explain in more detail why the dictionary values containing the children will also be deterministic?

Reply:
They're deterministic because all the traversals inside `group_memory_nodes` function are deterministic as well and based on `pipeline.nodes` order which is sorted already.

Reply:
> Can we add some comments about the intent of keeping it deterministic, it is not so obvious from the code. Ideally there should be tests for it but AFIAK it's a bit tricky to reproduce the randomness. This is only manually tested I assume.

I agree, impossible to test it

Reply:
> They're deterministic because all the traversals inside `group_memory_nodes` function are deterministic as well and based on `pipeline.nodes` order which is sorted already.

Is it means that previously code was not deterministic (and sorting was used) because of using `defaultdict`?

Reply:
Yes, exactly

Reply:
We tested manually, yes.

Theoretically, we can randomise `n` test cases and run each test `k` times to ensure we get the same results within runs. But given that there's no randomness, the results will be the same, so I don't think we need it.

Reply:
The randomness cannot be simulated with n use cases, this is another stories, but I agree we don't need it here.

Reply:
Randomness - no, but you can make a stress test for determinism, which will not guarantee it with 100% chance, but for large n and k, it will be close to it. Anyway, creating an adjacency list is deterministic, and `dfs` is deterministic; except for them, we have other loops which rely on `Pipeline` API, so there's no place from where randomness can come. So, I agree on not doing anything with it. 🙂